google/gax/bundling.py

```diff
@@ -0,0 +1,252 @@
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""Provides behavior that supports request bundling."""
+
+from __future__ import absolute_import
+
+import collections
+import threading
+
+
+def _str_dotted_getattr(obj, name):
+    """Expands extends getattr to allow dots in x to indicate nested objects.
+
+    Args:
+       obj: an object
+       name: a name for a field in the object
+
+    Returns:
+       the value of named attribute
+
+    Raises:
+       AttributeError if the named attribute does not exist
+    """
+    if name.find('.') == -1:
+        return getattr(obj, name)
+    for part in name.split('.'):
+        obj = getattr(obj, part)
+    return str(obj) if obj is not None else None
+
+
+def compute_bundle_id(obj, descriminator_fields):
+    """Computes a bundle id from the descriminator fields of `obj`.
+
+    descriminator_fields may include '.' as a separator, which is used to
+    indicate object traversal.  This is meant to allow fields in the
+    computed bundle_id.
+
+    the id is a tuple computed by going through the descriminator fields in
+    order and obtaining the str(value) object field (or nested object field)
+
+    if any descriminator field cannot be found, ValueError is raised.
+
+    Args:
+      obj: an object
+      descriminator_fields: a list of descriminator fields in the order to be
+        to be used in the id
+
+    Returns:
+      tuple: computed as described above
+
+    Raises:
+      AttributeError: if any descriminator fields attribute does not exist
+    """
+    return tuple(_str_dotted_getattr(obj, x) for x in descriminator_fields)
+
+
+class Task(object):
+    """Coordinates the execution of a single bundle."""
+
+    def __init__(self, api_call, bundle_id, bundled_field, bundling_request):
+        """Constructor.
+
+        Args:
+           api_call (callable[[object], object]): the func that is this tasks's
+             API call
+           bundle_id: the id of this bundle
+           bundled_field: the field used to create the bundled request
+           bundling_request: the request to pass as the arg to api_call
+
+        """
+        self._api_call = api_call
+        self._bundling_request = bundling_request
+        self.bundle_id = bundle_id
+        self.bundled_field = bundled_field
+        self.in_deque = collections.deque()
+        self.out_deque = collections.deque()
```

Comment:
I expect not, but I modeled the outputs using a deque in case we do need to allow for multiple responses when there are multiple bundle inputs.  if that's not necessary, I will eventually replace the output deque with something simpler.